src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,890 @@
+                                  callback);
+    }
+}
+
+TheISA::PCState
+SDCPUThread::getNextPC()
+{
+    if (inflightInsts.empty()) {
+        DPRINTF(SDCPUThreadEvent, "Getting PC value from committed state\n");
+        return _committedState->pcState();
+    } else {
+        const shared_ptr<InflightInst>& inst_ptr = inflightInsts.back();
+        DPRINTF(SDCPUThreadEvent, "Getting PC value from inst (seq %d)\n",
+                                  inst_ptr->seqNum());
+
+        assert(!inst_ptr->staticInst()->isControl() || inst_ptr->isComplete());
+
+        TheISA::PCState ret = inst_ptr->pcState();
+        if (!inst_ptr->staticInst()->isMacroop()) {
+            inst_ptr->staticInst()->advancePC(ret);
+        }
+        return ret;
+    }
+}
+
+ThreadContext*
+SDCPUThread::getThreadContext()
+{
+    return this;
+}
+
+void
+SDCPUThread::handleFault(std::shared_ptr<InflightInst> inst_ptr)
+{
+    DPRINTF(SDCPUThreadEvent, "Handling fault (seq %d): %s",
+                              inst_ptr->seqNum(), inst_ptr->fault()->name());
+
+    inflightInsts.clear();
+    // TODO may need to move any uncommitted InflightInsts somewhere else
+    //      to guarantee the memory is still valid until all events
+    //      involving those objects are handled.
+
+    inst_ptr->fault()->invoke(this, inst_ptr->staticInst());
+    decoder.reset();
+
+    advanceInst();
+}
+
+void
+SDCPUThread::issueInstruction(weak_ptr<InflightInst> inst)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr || inst_ptr->isSquashed()) {
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    DPRINTF(SDCPUInstEvent, "Issuing instruction (seq %d)\n",
+                            inst_ptr->seqNum());
+
+    populateDependencies(inst_ptr);
+    populateUses(inst_ptr);
+
+    if (inst_ptr->isReady()) { // If no dependencies, execute now
+        executeInstruction(inst);
+    } else { // Else, add an event callback to execute when ready
+        inst_ptr->addReadyCallback([this, inst](){
+                executeInstruction(inst);
+        });
+    }
+
+    if (inst_ptr->staticInst()->isLastMicroop()) {
+        curMacroOp = StaticInst::nullStaticInstPtr;
+    }
+
+    if (hasNextPC()) {
+        // If we are still running and know where to fetch, we should
+        // immediately send the necessary requests to the CPU to fetch the next
+        // instruction
+        advanceInst();
+    }
+}
+
+void
+SDCPUThread::markFault(shared_ptr<InflightInst> inst_ptr, Fault fault)
+{
+    DPRINTF(SDCPUThreadEvent, "Fault detected (seq %d): %s\n",
+                              inst_ptr->seqNum(), fault->name());
+
+    inst_ptr->fault(fault);
+
+    while (inflightInsts.back() != inst_ptr) {
+        inflightInsts.back()->squash();
+        inflightInsts.pop_back();
+    }
+
+    inst_ptr->squash();
+
+    commitAllAvailableInstructions();
```

Comment:
Hmm... this confuses me. Could you explain more what this is doing here?

Reply:
So the faulted instruction should be squashed, since the fault meant it couldn't make it safely to completion, and then we make the call to commitAllAvailableInstructions to basically check if inst_ptr is at the head of the inflightInsts table. commitAllAvailableInstructions() also handles faults (at commit time) so since that functionality is shared, I call the same instruction as when I say complete execution, as both events potentially produce commit-time jobs to do.